### ISSUE TITLE
Access Denied on ApplicationFolder if path contains Umlaute/Acentuated characters

### ISSUE THREAD

**Issue by creambyemute** [microsoft/react-native-windows#9315]
### Problem Description

We had a few weird bug reports from Users that were not able to login to the app at all. It hung on the Login-Screen.
After some investigation, it turned out that these Users all had an Umlaut or Acentuated Character in their Windows-Username and ran into an "Access Denied" when the App tried to create a folder within itself.

Funny enough, If you make this user an admin, it starts working? But that's definitely not a solution.

I also cannot tell our Customers to not use Umlaute or Acentuated Characters in their Username, French and Germans (and probably some more countries) will absolutely despise that advice.

I hope someone here has an idea what should be changed in order to make it work as my current attempts (u8path) didn't seem to work.

### Steps To Reproduce

1. Create a windows user with an Umlaut (ä, ö, ü etc) or an Acentuated character (é, à, è, ç etc) within the username
2. Init a new project
3. Add react-native-fs (https://github.com/wwimmo/react-native-fs)
4. `RNFS.mkdir(${RNFS.DocumentDirectoryPath} + "\someTestFolderToCreate")`

PS: It's not a problem if the folder to create contains an Umlaut or Acentuated character, but only if the Windows-Username folder contains an Umlaut/Acentuated character. Because it will try to create the user-directory with the Umlaut/Acentuated character inside `C:/Users/` instead of using the existing user-directory.

### Expected Results

Folder to be created, no access denied warning.

### CLI version

6.1.0

### Environment

```markdown
info Fetching system and libraries information...
System:
    OS: Windows 10 10.0.19043
    CPU: (12) x64 AMD Ryzen 5 5600X 6-Core Processor
    Memory: 21.91 GB / 31.94 GB
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.15 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK:
      API Levels: 28, 29, 30
      Build Tools: 29.0.0, 29.0.1, 29.0.2, 29.0.3, 30.0.0, 30.0.1, 30.0.2, 30.0.3
      System Images: android-28 | Google APIs Intel x86 Atom, android-29 | Google APIs Intel x86 Atom, android-30 | Google APIs Intel x86 Atom
      Android NDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.16299.0, 10.0.17134.0, 10.0.17763.0, 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.9.31313.79 (Visual Studio Community 2019)
  Languages:
    Java: 1.8.0_312 - C:\Program Files\OpenJDK\openjdk-8u312-b07\bin\javac.EXE
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2
    react-native: 0.65.2 => 0.65.2
    react-native-windows: 0.65.9 => 0.65.9
  npmGlobalPackages:
    *react-native*: Not Found
```

### Target Platform Version

10.0.19041

### Target Device(s)

Desktop

### Visual Studio Version

Visual Studio 2019

### Build Configuration

Release

### Snack, code example, screenshot, or link to a repository

I took this snippet from RNFS to demonstrate the Problem. When creating the `std::filesystem::path` it does some weird things with encoding it seems but `GetFolderFromPathAsync` is unable to handle that for the User-Folder? 

```c++
winrt::fire_and_forget RNFSManager::mkdir(std::string directory, RN::JSValueObject options, RN::ReactPromise<void> promise) noexcept
try
{
    size_t pathLength{ directory.length() };

    if (pathLength <= 0) {
        promise.Reject("Invalid path length");
    }
    else {
        bool hasTrailingSlash{ directory[pathLength - 1] == '\\' || directory[pathLength - 1] == '/' };
        // This makes the path unaccessible further down without admin rights
        // The umlaut/acentuated characters are being replaced when creating the path
        std::filesystem::path path(hasTrailingSlash ? directory.substr(0, pathLength - 1) : directory);
        path.make_preferred();

        auto parentPath{ path.parent_path().wstring() };
        std::stack<std::wstring> directoriesToMake;
        directoriesToMake.push(path.filename().wstring());

        StorageFolder folder{ nullptr };
        while (folder == nullptr) {
            try {
                // This fails with Access Denied altough it shouldn't
                folder = co_await StorageFolder::GetFolderFromPathAsync(parentPath);
            }
            catch (const hresult_error& ex) {
                hresult result{ ex.code() };
                if (result == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)) {
                    auto index{ parentPath.find_last_of('\\') };
                    directoriesToMake.push(parentPath.substr(index + 1));
                    parentPath = parentPath.substr(0, index);
                }
                else {
                    promise.Reject(winrt::to_string(ex.message()).c_str());
                }
            }
        }

        while (!directoriesToMake.empty()) {
            folder = co_await folder.CreateFolderAsync(directoriesToMake.top(), CreationCollisionOption::OpenIfExists);
            directoriesToMake.pop();
        }
        promise.Resolve();
    }
}
catch (const hresult_error& ex)
{
    // "Unexpected error while making directory."
    promise.Reject( winrt::to_string(ex.message()).c_str() );
}
```

**Comment by chrisglein** [comment]
> When creating the std::filesystem::path it does some weird things with encoding it seems but GetFolderFromPathAsync is unable to handle that for the User-Folder?

This may be due to a difference in expectations between what the std path uses (probably UTF 8?) and the UWP API expects (UTF 16).

There's not any React Native for Windows logic in the steps you're talking about. This should be logged in the RNFS repo and fixed there. Let me know if I'm not reading this correctly, or if you need help navigating the above.

**Comment by creambyemute** [comment]
@chrisglein I agree, it's a problem of RNFS and will need to be fixed in there.

I'm trying to find a solution to it and hoped someone here has an idea here on how this needs to be handled in the C++ code.

I've looked at the RNFS and react-native-fetch-blob windows implementations and they all just use the ```std::path``` so the problem will need to be fixed in both libraries.

**Comment by namrog84** [comment]
Our internal RNFS implementation (since RNFS didn't have a c++ impl before) was having the same issue. And was waiting on that C++ branch to test that out.
I think one of my coworkers fixed it today because we had a c_str() somewhere, and that was just switched to a string pass thru instead. Hopefully we can get a more thoroughly tested fix for everyone in that new c++ impl though.

Our internal use case was pretty limited so we didn't need to use the mkdir or getfolderfrompath flow. So the problem likely persists there.

**Comment by rozele** [comment]
cc @asklar - I think this is another issue where using fs::path instead of fs::path::u8path is problematic.

**Comment by asklar** [comment]
yep I believe it :) @rozele could you work with  @creambyemute to push a fix into @ronradtke 's fs module? https://github.com/RonRadtke/react-native-blob-util
